### ISSUE TITLE
Why must a Memento TimeMap be an LDP Container?

### ISSUE THREAD

**Issue by acoburn** [fcrepo/fcrepo-specification#12]

**Comment by ajs6f** [comment]
How did you infer  "`ldp:contains` triples are not allowed in the `LDPCv`"? In fact, the contained resources _are_ the very versions of the original resource that we want to maintain (the `LDPRm`s). This is troubling-- is the spec language really _that_ unclear... urg.

**Comment by ajs6f** [comment]
Maybe we need some pictures?

**Comment by acoburn** [comment]
The relevant part of the Memento RFC is: 

> A TimeMap is introduced to support retrieving a comprehensive list of all Mementos for a specific Original Resource known to a server.  The entity-body of a response to an HTTP GET request issued against a TimeMap's URI-T:

...

> MUST list the URI-M and archival datetime of each Memento for the Original Resource known to the server, preferably in a single document, or, alternatively in multiple documents that can be gathered by following contained links with a "timemap" Relation Type;

What I was thinking was this -- given a `URI-T`, serialized as `application/link-format`, the "contained" `Memento`s are all inlined such as (as per RFC 7089):

```
<http://repo.example.org/resource>;rel="original",
    <http://repo.example.org/resource?format=timemap>
      ; rel="self";type="application/link-format"
      ; from="Tue, 20 Jun 2000 18:02:59 GMT"
      ; until="Wed, 09 Apr 2008 20:30:51 GMT",
    <http://repo.example.org/resource>
      ; rel="timegate",
    <http://repo.example.org/resource?version=20000620180259>
      ; rel="first memento";datetime="Tue, 20 Jun 2000 18:02:59 GMT"
      ; license="http://creativecommons.org/publicdomain/zero/1.0/",
    <http://repo.example.org/resource?version=20091027204954>
       ; rel="last memento";datetime="Tue, 27 Oct 2009 20:49:54 GMT"
       ; license="http://creativecommons.org/publicdomain/zero/1.0/",
    <http://repo.example.org/resource?version=20000621011731>
      ; rel="memento";datetime="Wed, 21 Jun 2000 01:17:31 GMT"
      ; license="http://creativecommons.org/publicdomain/zero/1.0/",
    <http://repo.example.org/resource?version=20000621044156>
      ; rel="memento";datetime="Wed, 21 Jun 2000 04:41:56 GMT"
      ; license="http://creativecommons.org/publicdomain/zero/1.0/" .
```

Are you suggesting that an equivalent serialization of the URI-T would be:

```
<http://repo.example.org/resource?format=timemap>
    ldp:contains <http://repo.example.org/resource?version=20000620180259> ;
    ldp:contains <http://repo.example.org/resource?version=20091027204954> ;
    ldp:contains <http://repo.example.org/resource?version=20000621011731> ;
    ldp:contains <http://repo.example.org/resource?version=20000621044156> ;

    iana:first <http://repo.example.org/resource?version=20000620180259> ;
    iana:last <http://repo.example.org/resource?version=20091027204954> ;

    iana:memento <http://repo.example.org/resource?version=20000620180259> ;
    iana:memento <http://repo.example.org/resource?version=20091027204954> ;
    iana:memento <http://repo.example.org/resource?version=20000621011731> ;
    iana:memento <http://repo.example.org/resource?version=20000621044156> .
```

I'm not sure how using `ldp:contains` satisfies the "MUST list the URI-M and archival datetime ... in multiple documents that can be gathered by following contained links with a "timemap" Relation Type".

Mostly what I want to suss out is this: how would using LDP containment here improve simply using a simple LDP-RDFSource such as:

```
<http://repo.example.org/resource?format=timemap>
    dc:created "Tue, 20 Jun 2000 18:02:59 GMT" ;
    dc:modfied "Wed, 09 Apr 2008 20:30:51 GMT" ;
    iana:timemap <#v1>, <#v2>, <#v3>, <#v4> .
    iana:first <http://repo.example.org/resource?version=20000620180259> ;
    iana:last <http://repo.example.org/resource?version=20091027204954> ;

<#v1> iana:memento <http://repo.example.org/resource?version=20000620180259> ;
    dc:created "Tue, 20 Jun 2000 18:02:59 GMT" .

<#v2> iana:memento <http://repo.example.org/resource?version=20000621011731> ;
    dc:created "Wed, 21 Jun 2000 01:17:31 GMT" .

<#v3> iana:memento <http://repo.example.org/resource?version=20000621044156> ;
    dc:created "Wed, 21 Jun 2000 04:41:56 GMT" .

<#v4> iana:memento <http://repo.example.org/resource?version=20091027204954> ;
    dc:created "Tue, 27 Oct 2009 20:49:54 GMT" .
```

And if the `LDPCv` is understood to be a LDP Container, how does the Fedora spec reconcile LDP membership triples with the triples in a TimeMap? If some other resource points to the `LDPCv` as a `ldp:membershipResource`, must _those_ triples become part of the timemap?

**Comment by acoburn** [comment]
My basic confusion here has to do with the RFC 7089 recommendation that a TimeMap should be a single document, but that it can be spread across multiple documents by following the iana:timemap property -- i.e. if the other documents are _also_ TimeMaps, not just `Memento`s. But if Fedora is suggesting using `ldp:contains` to link to `Memento`s, unless I am misunderstanding the semantics of `ldp:contains`, the Fedora guidance seems to be recommending a very different model of assembling the data in a TimeMap.

**Comment by ajs6f** [comment]
(If it makes you feel any better about this, although it is an appeal to authority, @azaroth42 actively liked this approach, @hvdsomp okayed it with some enthusiasm after discussion on the memento-dev@ list, and @phonedude found it unobjectionable when I asked him about it in person. They wrote RFC7089. If they don't know Memento, no one does.)

This isn't an either/or situation. The `LDPCv` is _both_ an `LDPC` _and_ a `Timemap`. The links between it and the `LDPRm`s are _both_ `ldp:contains` _and_ `iana:memento`. Why do we want it to be an `LDPC`? Two immediate reasons:
1. So that `LDP` clients that don't know Fedora or Timemap can find the versions of things by following ordinary LDP patterns. Remember, LDP is the "ethos" of the Fedora API. Containment is the natural idiom in LDP that is closest to the relationship between a `Timemap` and its `URI-M`s.
2. [This](https://fcrepo.github.io/fcrepo-specification/index.html#h-client-managed) is what supports important use cases from @barmintor and @azaroth42, including the provision of versions that weren't created via snapshot in the Fedora API (e.g. to migrate from other systems or for export/import cycling a la what @ruebot and crew are working on).

That second point brings us back to the most important idea in the versioning spec:
1. The Fedora API is RESTful. =>
2. It deals in representations, not directly in resources. =>
3. It does not claim to deal in the _entire_ history of a resource, but in selected versions ( = snapshot representations) of that resource. =>
4. It must be able to handle versions it did not create and history it does not "perceive". (It's _not_ a VCS such as we use for code, it's an API to arbitrary implementations.)

That supports the use cases in point 2 above and necessitates some way to provide versions that weren't created as snapshots of resources in the repository (e.g. in the migration case, snapshots of resources that are _now_ in the repository but weren't always), which is the API section in point 2 above.

As for the multiple-docs-Timemap question-- the spec is implicitly offering a particular topology for the "preferably"/"or" of the RFC. We are saying "don't use chained `Timemap`s, use multiple `Timemaps`/`LDPCv`s linked directly to the `URI-R`/`LDPRv`" (which is perfect legal according to both the RFC and the spec). Maybe that needs to be made explicit?

Another way to put this relying more on constraint:
We have to offer some means to create versions. But because of the logic in the second list above and the use cases, we also have to offer some means to accept versions we didn't create. The first demand we could satisfy just with action against the versioned resource. For the second, that would be pretty unnatural. (What could you do, `PUT` a bitstream against the original resource with a header set so that instead of actually changing the state of that resource, it creates a different resource somewhere else? Icky, to say the least.) What is more @hvdsomp and the RFC explicitly remark that the original resource cannot be its own `Timegate`, but that the `Timegate` and the `Timemap` can be the same. So we have to have another resource against which to act. Everything points to it being the  `Timegate`/`Timemap`. So how do we act against the  `Timegate`/`Timemap` to create resources? This is where we bring LDP back in to govern that interaction. We don't _have_ to do that. We could say instead that "Everything in the repo is LDP, _except_ these certain resources which have a completely different interaction model, one that is unique to Fedora." That is not an attractive alternative. 

Does that help?

**Comment by ajs6f** [comment]
TLDR; what LDP containment offers us over your examples is not a way to write down relationships but a way to _create_ them. On a basis of representation only, your examples are just as good as the spec. It is the interaction models that the spec is bringing via LDP that make it useful. And we are writing an API, so we have to offer both representation _and_ interaction.

**Comment by acoburn** [comment]
@ajs6f yes, that is very helpful. One point of clarification, though: in RDF 7089 [Section 4.1](https://tools.ietf.org/html/rfc7089#section-4) "The Original Resource Acts as Its Own TimeGate" is explicitly called out as an implementation pattern, so I'm not sure how that squares with your statement that "the original resource cannot be its own `Timegate`".

**Comment by hvdsomp** [comment]
On Dec 27, 2016, at 14:28, Aaron Coburn <notifications@github.com> wrote:
> 
> @ajs6f yes, that is very helpful. One point of clarification, though: in RDF 7089 Section 4.1 "The Original Resource Acts as Its Own TimeGate" is explicitly called out as an implementation pattern, so I'm not sure how that squares with your statement that "the original resource cannot be its own Timegate".
> 

Indeed. Original Resource can coincide with its TimeGate. But TimeGate and TimeMap must be different. 

Herbert

**Comment by ajs6f** [comment]
Of course, you are quite right! I mis-wrote there. (Too early and too little tea before writing!)

**Comment by acoburn** [comment]
OK, that's good. I'm satisfied with that and will close this issue. Thanks for providing all of the context!

**Comment by ajs6f** [comment]
In fact, I think the correct formulation makes the argument stronger, if anything. I will leave it wrong to preserve history on the assumption that anyone reading this will read down to the bottom.
